network/framework/src/application/storage.rs

```diff
@@ -188,21 +201,39 @@ impl PeersAndMetadata {
             .and_modify(|peer_metadata| {
                 peer_metadata.connection_metadata = connection_metadata.clone()
             })
-            .or_insert_with(|| PeerMetadata::new(connection_metadata));
+            .or_insert_with(|| PeerMetadata::new(connection_metadata.clone()));
 
         // Update the cached peers and metadata
         self.set_cached_peers_and_metadata(peers_and_metadata.clone());
 
+        let net_context = self.get_network_context(&peer_network_id.network_id());
+        let event = ConnectionNotification::NewPeer(connection_metadata, net_context);
+        self.broadcast(event);
+
         Ok(())
     }
 
+    pub fn set_network_contexts(&self, contexts: BTreeMap<NetworkId, NetworkContext>) {
+        self.contexts.store(Arc::new(contexts));
+    }
+
+    // Return NetworkContext for a NetworkId, or fake-up something best effort because it's just for logging.
+    fn get_network_context(&self, network_id: &NetworkId) -> NetworkContext {
```

Comment:
Hmm... It might be cleaner to make this return None as opposed to an arbitrary context 😄

Reply:
It should never miss, I'm just chicken about putting in a panic!() or unreachable!()
This gets set up in node `setup_networks_and_get_interfaces()` based on the node config, and if somehow the system later gets a message on a network it didn't setup at start time, something very weird has happened

Reply:
> I'm just chicken about putting in a panic!() or unreachable!()

Gotcha. Why not `Option`, though? These types of cases are exactly what options are intended for? 😄 

> something very weird has happened

Yeah, this is why I think we shouldn't fail silently. Returning an arbitrary value feels... weird.

Reply:
returning Option is good. DOne.